ZNTA-2057 async refactor

huangp opened the pull request:
JIRA issue URL: https://zanata.atlassian.net/browse/ZNTA-2057

* refactor async key from Serializable to Object
* added REST endpoint for querying async tasks (http://localhost:8080/rest/process?includeFinished=true)


## Reviewer tasks

The following is based on the
[Zanata Development Guidelines](https://github.com/zanata/zanata-platform/wiki/Development-Guidelines).
As a reviewer, you should check the guidelines regularly for updates, and just
to refresh your memory.


The reviewer can use this list for reference to make sure
they have considered all the important points.

Use the checkboxes or not, whatever works best for you.

 - Code quality
   - [ ] Code passes style check (checkstyle/eslint)
   - [ ] Code is clear and concise
   - [ ] UI code is internationalised
   - [ ] Code has adequate comments where appropriate
   - Code is in an appropriate place
     - [ ] Classes are in appropriate packages
     - [ ] Code fits with class's responsibilities (SRP)
   - [ ] Configuration files are documented enough
   - If code is removed
     - [ ] No remaining code references the removed code
       - [ ] No orphaned rewrite rules
       - [ ] No orphaned config settings
     - [ ] No remaining comments refer to the removed code
     - [ ] Unused imports and libraries are removed
 - Testing
   - [ ] New code has tests
   - [ ] Changed code has tests
   - [ ] All tests pass
   - [ ] Test coverage stable or increased
 - Documentation
   - [ ] New features are documented
   - [ ] Docs removed for deleted features
   - [ ] Docs updated for all changed features
   - [ ] Developer/sysadmin docs updated if appropriate
 - If there are new dependencies
   - [ ] Does each new dependency pass all the
         [Considerations for new dependencies](https://github.com/zanata/zanata-platform/wiki/Development-Guidelines#new-technologies-and-dependencies)?


----
*This template can be updated in .github/PULL_REQUEST_TEMPLATE.md*

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/zanata/zanata-platform/398)
<!-- Reviewable:end -->

seanf commented:
Reviewed 25 of 26 files at r1.
Review status: 25 of 26 files reviewed at latest revision, 16 unresolved discussions.

---

*[server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java, line 94 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-KmxzMm_6Ler0WSxRn1a:-KmxzMmaXlcRG9rbf-5K:b-291wtf) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java#L94)):*
> ```Java
>             return handlesByKey.get(key);
>         }
>         return finishedTasks.getIfPresent(key.id());
> ```

Can we use the Key itself as a, well, key? Might help with the O(n) problems in this class.

We might also want to consider having a single key class (various factory methods, but just one class) so that it's easy to turn an id string back into a Key object.

---

*[server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java, line 139 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-KmxyBlUGdsRP9NefuFq:-KmxyBlUGdsRP9NefuFr:b-9i2afj) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java#L139)):*
> ```Java
>         String SEPARATOR = "-";
> 
>         String id();
> ```

It might be worth adding a default implementation for toString, which could call id.

---

*[server/zanata-war/src/main/java/org/zanata/async/UserTriggerableTaskHandle.java, line 23 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy1b7HzmOztRQa5hQg:-Kmy1b7IZnCQCDnPxb8k:bn03v6g) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/async/UserTriggerableTaskHandle.java#L23)):*
> ```Java
> package org.zanata.async;
> 
> public interface UserTriggerableTaskHandle {
> ```

Seems more like CancellableTask?

---

*[server/zanata-war/src/main/java/org/zanata/async/handle/TransMemoryMergeTaskHandle.java, line 61 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-KmxxXlFcFsfZl0Kk4q-:-KmxxXlFcFsfZl0Kk4q0:b-t8j750) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/async/handle/TransMemoryMergeTaskHandle.java#L61)):*
> ```Java
>                 .add("mergeTarget", mergeTarget)
>                 .add("textFlowFilled", currentProgress)
>                 .add("totalTextFlows", maxProgress)
> ```

These changes make it unclear what currentProgress/maxProgress mean for this type of task. Perhaps we should keep around the old getters/setters and have them delegate to get/set currentProgress/maxProgress. (Or some other solution.)

---

*[server/zanata-war/src/main/java/org/zanata/async/handle/TransMemoryMergeTaskHandle.java, line 62 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-KmxxGKfH7nR1C9JwRzR:-KmxxGKgcbsnLFEA0V3e:b5rm3mw) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/async/handle/TransMemoryMergeTaskHandle.java#L62)):*
> ```Java
>                 .add("textFlowFilled", currentProgress)
>                 .add("totalTextFlows", maxProgress)
>                 .add("cancelledBy", getTriggeredBy())
> ```

This assumes that the canceller is always the user who triggered the task. This will be wrong if we let admin cancel user tasks. Perhaps you should add a method getCancelledBy() which calls getTriggeredBy() for now.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsynchronousProcessResourceService.java, line 117 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy-SWTiX27AkcauivW:-Kmy-SWTiX27AkcauivX:be1vn17) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsynchronousProcessResourceService.java#L117)):*
> ```Java
>                         resource, extensions, copytrans, true, handle);
>         logWhenUploadComplete(handle, name, taskKey.id());
>         return getProcessStatus(taskKey.id()); // TODO Change to return 202
> ```

Why not pass in the key itself to these methods (and to AsyncTaskHandleManager)?

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 100 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy36qbzz-njmfomdK_:-Kmy36qbzz-njmfomdKa:b-e4vkl8) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L100)):*
> ```Java
>         ProcessStatus status = handleToProcessStatus(handle,
>                 uriInfo.getRequestUri().toString());
>         return Response.ok(status).build();
> ```

We should do an authorisation check.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 104 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy09TlxYGtm5zyTHSb:-Kmy09TlxYGtm5zyTHSc:bnqktim) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L104)):*
> ```Java
> 
>     /**
>      * Get all async task's status.
> ```

Get all async task's status. -> Get all async tasks' statuses. (or Get status of/for all async tasks.)

Is this for all users, or just the current user? If the latter, seems like an admin-only function.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 118 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy0dolz5KAikFua7Le:-Kmy0dolz5KAikFua7Lf:b-c2f4k6) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L118)):*
> ```Java
>      */
>     @GET
>     public Response getAllAsyncProcess(
> ```

The other method has "status" in the name.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 155 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy1M9tQF-muxil6xnt:-Kmy1M9tQF-muxil6xnu:baaxfd7) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L155)):*
> ```Java
>             return Response.status(Response.Status.NOT_FOUND).build();
>         }
>         if (handle.isStarted() && !handle.isCancelled() && !handle.isDone()) {
> ```

Can we encapsulate that boolean logic into a single method somewhere?

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 159 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy1kUkGV8xWKOx4cvC:-Kmy1kUkGV8xWKOx4cvD:bl1xgph) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L159)):*
> ```Java
>                 UserTriggerableTaskHandle taskHandle =
>                         (UserTriggerableTaskHandle) handle;
>                 if (!taskHandle.canCancel(identity.getAccountUsername())) {
> ```

We should try to encapsulate all that logic into the canCancel method.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 161 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy0xNapHSVUleDA553:-Kmy0xNapHSVUleDA554:b8pn4yt) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L161)):*
> ```Java
>                 if (!taskHandle.canCancel(identity.getAccountUsername())) {
>                     throw new AuthorizationException(
>                             "Only the task triggerer or admin can cancel the task.");
> ```

"Triggerer" sounds a bit unwieldy. Perhaps "owner" or "creator" would sound better?

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 165 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy27J29PqVu2sL-cCI:-Kmy27J29PqVu2sL-cCJ:bb3od56) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L165)):*
> ```Java
>             }
> 
>             handle.cancel(true);
> ```

Does this mean that some types of task can be cancelled by *any* authenticated user? That doesn't seem right.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 182 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy2pXqp60pfJ1TA7Cz:-Kmy2pXqp60pfJ1TA7D-:bcr8its) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L182)):*
> ```Java
>                 handle.isDone() ? ProcessStatus.ProcessStatusCode.Finished
>                         : ProcessStatus.ProcessStatusCode.Running);
>         int perComplete = 100;
> ```

percentComplete would be better. Or just percentage.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 191 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy3MwzSuAIr5W9twcg:-Kmy3MwzSuAIr5W9twch:b-owod2t) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L191)):*
> ```Java
>         if (handle.isCancelled()) {
>             status.setStatusCode(ProcessStatus.ProcessStatusCode.Cancelled);
>             status.addMessage("Cancelled");
> ```

Do we want to include the username?

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 198 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy3G88attISjLcQYuc:-Kmy3G88attISjLcQYud:b4qn7mx) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L198)):*
> ```Java
>             } catch (InterruptedException e) {
>                 // The process was forcefully cancelled
>                 status.setStatusCode(ProcessStatus.ProcessStatusCode.Failed);
> ```

Shouldn't this say ProcessStatusCode.Cancelled then? Do we want to include the username?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/398)*
<!-- Sent from Reviewable.io -->

huangp commented:
Review status: 22 of 26 files reviewed at latest revision, 16 unresolved discussions.

---

*[server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java, line 94 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-KmxzMm_6Ler0WSxRn1a:-Kmy8AoD7Qx3ehWdA4xB:b8jbjny) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java#L94)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Can we use the Key itself as a, well, key? Might help with the O(n) problems in this class.

We might also want to consider having a single key class (various factory methods, but just one class) so that it's easy to turn an id string back into a Key object.
</blockquote></details>

Fixed. In my first commit I actually attempted to add methods to convert between id and object. But then I realize there is no use case to convert from id to key object (other than saving us from the O(n). Given that we have a thread pool of fixed size so that n is not going to be huge. I didn't bother going that route. 

---

*[server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java, line 139 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-KmxyBlUGdsRP9NefuFq:-Kmy7mJNqyAqPPWKyoh4:byrzy4v) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java#L139)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

It might be worth adding a default implementation for toString, which could call id.
</blockquote></details>

We can't do that. Java won't allow it.

---

*[server/zanata-war/src/main/java/org/zanata/async/UserTriggerableTaskHandle.java, line 23 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy1b7HzmOztRQa5hQg:-Kmy8rDyXGltcLlJdAHR:bc87usz) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/async/UserTriggerableTaskHandle.java#L23)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Seems more like CancellableTask?
</blockquote></details>

Well all tasks can be cancelled (by admin). Is just some tasks are triggered by user and should only be cancelled by that user. 

---

*[server/zanata-war/src/main/java/org/zanata/async/handle/TransMemoryMergeTaskHandle.java, line 61 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-KmxxXlFcFsfZl0Kk4q-:-Kmy9E-83lXfSe4HRTUi:b-gogrrr) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/async/handle/TransMemoryMergeTaskHandle.java#L61)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

These changes make it unclear what currentProgress/maxProgress mean for this type of task. Perhaps we should keep around the old getters/setters and have them delegate to get/set currentProgress/maxProgress. (Or some other solution.)
</blockquote></details>

When I implemented this I didn't realize there are progress fields in the super class. I could just reuse those two fields. The display didn't mention the field names (just a percent), so it's fine to just remove the two fields and use progress fields. I will rename the toString method (only used in debug I believe)

---

*[server/zanata-war/src/main/java/org/zanata/async/handle/TransMemoryMergeTaskHandle.java, line 62 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-KmxxGKfH7nR1C9JwRzR:-Kmy9oXW6ZSBg0eagYiI:bqvohh4) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/async/handle/TransMemoryMergeTaskHandle.java#L62)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

This assumes that the canceller is always the user who triggered the task. This will be wrong if we let admin cancel user tasks. Perhaps you should add a method getCancelledBy() which calls getTriggeredBy() for now.
</blockquote></details>

Right. In fact the toString is only used in log or debugging. I will get rid of it so It's not confusing. 

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsynchronousProcessResourceService.java, line 117 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy-SWTiX27AkcauivW:-KmyAKTvAl6bfnHAF7Ir:b1wfhaz) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsynchronousProcessResourceService.java#L117)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Why not pass in the key itself to these methods (and to AsyncTaskHandleManager)?
</blockquote></details>

getProcessStatus is a method in API. Until we deprecate the method (move this to the generic Async REST API), we can't change it.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 100 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy36qbzz-njmfomdK_:-KmyB4zblbWpu7uY2e9k:bbmz7jh) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L100)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

We should do an authorisation check.
</blockquote></details>

Only admin and the task owner can see the process status? Do we need to be that strict for get GET?

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 104 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy09TlxYGtm5zyTHSb:-KmyBW2_denJGS3RUMwk:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L104)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Get all async task's status. -> Get all async tasks' statuses. (or Get status of/for all async tasks.)

Is this for all users, or just the current user? If the latter, seems like an admin-only function.
</blockquote></details>

Done.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 118 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy0dolz5KAikFua7Le:-KmyBi5792A-ZUVjtueo:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L118)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

The other method has "status" in the name.
</blockquote></details>

Done.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 155 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy1M9tQF-muxil6xnt:-Kmy546l399ZBLYEUaqF:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L155)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Can we encapsulate that boolean logic into a single method somewhere?
</blockquote></details>

Done.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 159 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy1kUkGV8xWKOx4cvC:-Kmy4Fnrv45tbiEJj225:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L159)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

We should try to encapsulate all that logic into the canCancel method.
</blockquote></details>

Done.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 161 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy0xNapHSVUleDA553:-Kmy4Xpm-H4cv3q5YHEY:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L161)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

"Triggerer" sounds a bit unwieldy. Perhaps "owner" or "creator" would sound better?
</blockquote></details>

Done.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 165 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy27J29PqVu2sL-cCI:-Kmy4dlyrrDqbJ4g_qXa:bncggmn) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L165)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Does this mean that some types of task can be cancelled by *any* authenticated user? That doesn't seem right.
</blockquote></details>

I forgot to implement the other part. For those "system" type task, we should only allow admin to cancel them.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 182 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy2pXqp60pfJ1TA7Cz:-Kmy6_y5RNduXtFCIfOk:b-csy3vc) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L182)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

percentComplete would be better. Or just percentage.
</blockquote></details>

This chunk was moved from other place. Will rename it.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 191 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy3MwzSuAIr5W9twcg:-Kmy6pGqd_eMogcwgboL:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L191)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Do we want to include the username?
</blockquote></details>

Done.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 198 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy3G88attISjLcQYuc:-Kmy6t2dgoRClgzcOA5w:b-xc49jz) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L198)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Shouldn't this say ProcessStatusCode.Cancelled then? Do we want to include the username?
</blockquote></details>

I moved this code from AsynchronousProcessResourceService. I found that if you cancel a task, upon handle.getResult(), it will throw a CancellationException instead of InterruptedException. That's why I added the above cancel check block. I am not sure whether this block will even work. Maybe only if you can somehow get hold of the running thread and interrupt that thread (e.g. during a server shutdown?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/398)*
<!-- Sent from Reviewable.io -->

seanf commented:
Reviewed 6 of 6 files at r2.
Review status: 24 of 27 files reviewed at latest revision, 9 unresolved discussions.

---

*[server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java, line 94 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-KmxzMm_6Ler0WSxRn1a:-KmypZ6HG_joCqOkr4UR:b-ws49gj) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java#L94)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

Fixed. In my first commit I actually attempted to add methods to convert between id and object. But then I realize there is no use case to convert from id to key object (other than saving us from the O(n). Given that we have a thread pool of fixed size so that n is not going to be huge. I didn't bother going that route. 
</blockquote></details>

O(n) is actually pretty terrible, and getting rid of it is a valid use case. Just because the thread pool is limited doesn't really guarantee that n is small.

The only reason we need the O(n) methods is because we keep switching between these String ids and the map key. It's a sign that we are misusing our collections.

This method is called by the CLI (via AsynchronousProcessResource.getProcessStatus) fairly frequently when waiting for an upload to finish, so it's worth choosing good data structures.

If the key string ids are unique anyway, perhaps we should just key the hashmap by string id instead of the so-called Key object.

---

*[server/zanata-war/src/main/java/org/zanata/async/UserTriggerableTaskHandle.java, line 23 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy1b7HzmOztRQa5hQg:-KmylZx3j20ko_MrGSJh:b-m5v3l2) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/async/UserTriggerableTaskHandle.java#L23)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

Well all tasks can be cancelled (by admin). Is just some tasks are triggered by user and should only be cancelled by that user. 
</blockquote></details>

In that case it could probably use more javadoc, because it's still not clear what this really represents.

The reason I suggested that name is that the only method on this interface is about cancelling.

What about UserTriggeredTask? (Perhaps with Handle at the end.)

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsynchronousProcessResourceService.java, line 117 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy-SWTiX27AkcauivW:-KmyrAErIzAIlFxmhoo2:bmhgf3k) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsynchronousProcessResourceService.java#L117)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

getProcessStatus is a method in API. Until we deprecate the method (move this to the generic Async REST API), we can't change it.
</blockquote></details>

Oh, you mean it's in zanata-common-api? We have a lot of API besides zanata-common-api.

Why can't we deprecate it in place and add another? But I'm guessing the various handle/key classes aren't part of zanata-common-api. That's fine.

We could still add an overload here (not declared in zanata-common-api) which can accept a key object instead of the string. And logWhenUploadComplete can take a key object without changing the public API.

Remember that the current getProcessStatus(String) is currently O(n), but getProcessStatus(Key) would be O(1) - unless we switch that hashmap to be keyed by String of course. 

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 100 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy36qbzz-njmfomdK_:-Kmyqb4w_QVtuSsH4iTv:bqqm6gy) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L100)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

Only admin and the task owner can see the process status? Do we need to be that strict for get GET?
</blockquote></details>

Some level of strictness in between might be nice, but exposing any handle to any authenticated user (eg regardless of project affiliation) seems like a problem.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 198 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy3G88attISjLcQYuc:-KmyojBHV7sycMg6BeU2:be5j6fn) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L198)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

I moved this code from AsynchronousProcessResourceService. I found that if you cancel a task, upon handle.getResult(), it will throw a CancellationException instead of InterruptedException. That's why I added the above cancel check block. I am not sure whether this block will even work. Maybe only if you can somehow get hold of the running thread and interrupt that thread (e.g. during a server shutdown?)
</blockquote></details>

If we want preemptive cancellation, and at some point I think we will, we will need to make more use of InterruptedException in future. It's the best way to cancel a lot of things.

From an external point of view, whether it was "cancelled" or "interrupted" is an implementation detail which shouldn't be visible. If we're going to have a state for Cancelled (and we should), it should be used whenever a process is cancelled, preemptively or not.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 202 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmynr2BhPu3cN8SJEzQ:-Kmynr2CieNID3qTTohl:b8t1bnf) ([raw file](https://github.com/zanata/zanata-platform/blob/ddc82d3fc6a6c5c60144faa09e7fc5243bdbf23e/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L202)):*
> ```Java
>         handle.cancel(true);
>         handle.setCancelledBy(identity.getAccountUsername());
>         handle.setCancelledTime(System.currentTimeMillis());
> ```

Perhaps we should move all the canCancel check, user/admin cancel operation and setCancelled* calls into the handle.cancel method. We should be able to avoid the `if instanceof UserTriggerableTaskHandle` with a bit of polymorphism.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 211 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmymh1xvTkitElzlJnf:-Kmymh1y-5p3D24hIRa4:b-tzx6er) ([raw file](https://github.com/zanata/zanata-platform/blob/ddc82d3fc6a6c5c60144faa09e7fc5243bdbf23e/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L211)):*
> ```Java
>     private static boolean taskIsRunning(AsyncTaskHandle handle) {
>         return handle.isStarted() && !handle.isCancelled() && !handle.isDone();
>     }
> ```

This might be better off in the class AsyncTaskHandle.

---

*[server/zanata-war/src/test/java/org/zanata/rest/service/AsyncProcessServiceTest.java, line 88 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmytvgmfbii2WMF919q:-KmytvgnYX-nBYgCegnP:b-a9071) ([raw file](https://github.com/zanata/zanata-platform/blob/ddc82d3fc6a6c5c60144faa09e7fc5243bdbf23e/server/zanata-war/src/test/java/org/zanata/rest/service/AsyncProcessServiceTest.java#L88)):*
> ```Java
>         Response response = service.getAsyncProcessStatus("id");
> 
>         assertThat(response.getStatus()).isEqualTo(200);
> ```

We should test a user trying to access someone else's task too. Probably 404.

---

*[server/zanata-war/src/test/java/org/zanata/rest/service/AsyncProcessServiceTest.java, line 131 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-KmytegQ8RDNZfZ9SOSH:-KmytegQ8RDNZfZ9SOSI:budupjw) ([raw file](https://github.com/zanata/zanata-platform/blob/ddc82d3fc6a6c5c60144faa09e7fc5243bdbf23e/server/zanata-war/src/test/java/org/zanata/rest/service/AsyncProcessServiceTest.java#L131)):*
> ```Java
>         Response response = service.getAllAsyncProcessStatuses(true);
> 
>         assertThat(response.getStatus()).isEqualTo(200);
> ```

We should test a non-admin user for this end-point, and the one above. Probably UNAUTHORIZED.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/398)*
<!-- Sent from Reviewable.io -->

seanf commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/398)*
<!-- Sent from Reviewable.io -->

huangp commented:
Review status: 10 of 28 files reviewed at latest revision, 8 unresolved discussions.

---

*[server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java, line 94 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-KmxzMm_6Ler0WSxRn1a:-Kmz2vEi4sxjvkcETQo4:b-8rf7t5) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java#L94)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

O(n) is actually pretty terrible, and getting rid of it is a valid use case. Just because the thread pool is limited doesn't really guarantee that n is small.

The only reason we need the O(n) methods is because we keep switching between these String ids and the map key. It's a sign that we are misusing our collections.

This method is called by the CLI (via AsynchronousProcessResource.getProcessStatus) fairly frequently when waiting for an upload to finish, so it's worth choosing good data structures.

If the key string ids are unique anyway, perhaps we should just key the hashmap by string id instead of the so-called Key object.
</blockquote></details>

Done. I think the reason we chose Key object is because we can can have meaningful equal and hashcode implementation as well as toString method. Some tasks are meant to be run at most one at a time for same key (e.g. copy version). Since the id is generated from the field values, I think it's fine to use it as key.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsynchronousProcessResourceService.java, line 117 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy-SWTiX27AkcauivW:-Kmz4eljSr4jAPlIlfwE:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsynchronousProcessResourceService.java#L117)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Oh, you mean it's in zanata-common-api? We have a lot of API besides zanata-common-api.

Why can't we deprecate it in place and add another? But I'm guessing the various handle/key classes aren't part of zanata-common-api. That's fine.

We could still add an overload here (not declared in zanata-common-api) which can accept a key object instead of the string. And logWhenUploadComplete can take a key object without changing the public API.

Remember that the current getProcessStatus(String) is currently O(n), but getProcessStatus(Key) would be O(1) - unless we switch that hashmap to be keyed by String of course. 
</blockquote></details>

Done.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 100 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy36qbzz-njmfomdK_:-Kmz9H0aiVBjEHYP_zx2:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L100)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Some level of strictness in between might be nice, but exposing any handle to any authenticated user (eg regardless of project affiliation) seems like a problem.
</blockquote></details>

Done.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 198 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy3G88attISjLcQYuc:-KmzAn3kv6F7fDnDExTX:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L198)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

If we want preemptive cancellation, and at some point I think we will, we will need to make more use of InterruptedException in future. It's the best way to cancel a lot of things.

From an external point of view, whether it was "cancelled" or "interrupted" is an implementation detail which shouldn't be visible. If we're going to have a state for Cancelled (and we should), it should be used whenever a process is cancelled, preemptively or not.
</blockquote></details>

Done.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 202 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmynr2BhPu3cN8SJEzQ:-Kmz9Fbub5ElYkPb5FQL:b-e1mij6) ([raw file](https://github.com/zanata/zanata-platform/blob/ddc82d3fc6a6c5c60144faa09e7fc5243bdbf23e/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L202)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Perhaps we should move all the canCancel check, user/admin cancel operation and setCancelled* calls into the handle.cancel method. We should be able to avoid the `if instanceof UserTriggerableTaskHandle` with a bit of polymorphism.
</blockquote></details>

There are some handles that just call the cancel method without setting CancelledBy and CancelledTime. I tried not to make too big a change to the existing code.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 211 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmymh1xvTkitElzlJnf:-Kmz74r0Yf4SK-VXzKLr:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/ddc82d3fc6a6c5c60144faa09e7fc5243bdbf23e/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L211)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

This might be better off in the class AsyncTaskHandle.
</blockquote></details>

Done.

---

*[server/zanata-war/src/test/java/org/zanata/rest/service/AsyncProcessServiceTest.java, line 88 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmytvgmfbii2WMF919q:-Kn1iiEsdoYUPNQ0BBgq:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/ddc82d3fc6a6c5c60144faa09e7fc5243bdbf23e/server/zanata-war/src/test/java/org/zanata/rest/service/AsyncProcessServiceTest.java#L88)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

We should test a user trying to access someone else's task too. Probably 404.
</blockquote></details>

Done.

---

*[server/zanata-war/src/test/java/org/zanata/rest/service/AsyncProcessServiceTest.java, line 131 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-KmytegQ8RDNZfZ9SOSH:-Kn1xTCvJ7IkgkjK4aLo:bbrrjx8) ([raw file](https://github.com/zanata/zanata-platform/blob/ddc82d3fc6a6c5c60144faa09e7fc5243bdbf23e/server/zanata-war/src/test/java/org/zanata/rest/service/AsyncProcessServiceTest.java#L131)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

We should test a non-admin user for this end-point, and the one above. Probably UNAUTHORIZED.
</blockquote></details>

This is controlled by CDI interceptor. I added an arquillian test for this.

---

*[server/zanata-war/src/main/java/org/zanata/async/UserTriggerableTaskHandle.java, line 23 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmy1b7HzmOztRQa5hQg:-Kmz4XusOoiEjlKGNDbO:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/82fe345ae6382062a0439d8b4adf189196de57ca/server/zanata-war/src/main/java/org/zanata/async/UserTriggerableTaskHandle.java#L23)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

In that case it could probably use more javadoc, because it's still not clear what this really represents.

The reason I suggested that name is that the only method on this interface is about cancelling.

What about UserTriggeredTask? (Perhaps with Handle at the end.)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/398)*
<!-- Sent from Reviewable.io -->

seanf commented:
Reviewed 13 of 18 files at r5, 6 of 6 files at r6.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java, line 62 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kn1rMojkpDZMr806pgg:-Kn1rMojkpDZMr806pgh:bsa607z) ([raw file](https://github.com/zanata/zanata-platform/blob/e753e86ce8476996501daa1587731bee69e8a585/server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java#L62)):*
> ```Java
>                     + " already exists");
>         }
>         handlesByKey.put(key.id(), handle);
> ```

This could use HashMap.putIfAbsent now, instead of containsKey/put.

---

*[server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java, line 72 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kn1s9xbDtOjGQM2qVoM:-Kn1s9xbDtOjGQM2qVoN:bsakcy1) ([raw file](https://github.com/zanata/zanata-platform/blob/e753e86ce8476996501daa1587731bee69e8a585/server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java#L72)):*
> ```Java
>     public synchronized String registerTaskHandle(AsyncTaskHandle handle) {
>         String autoGenKey = UUID.randomUUID().toString();
>         AsyncTaskKey autoKey = () -> autoGenKey;
> ```

So we're using AsyncTaskKey like a FunctionalInterface (even though it doesn't have the annotation)? This seems error-prone, because the temptation in this method would be to write:

    AsyncTaskKey autoKey = () -> UUID.randomUUID().toString();

Whereas one of the most important things about a key is that it be immutable.

If it's going to be a FunctionalInterface, perhaps it would be better as an AsyncTaskKeyGen (after all, it just generates the id string which is the real key now).

Otherwise it might be safest if AsyncTaskKey were a final immutable class which just holds a String.

---

*[server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java, line 82 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kn1uSp9mOSKU4Rq2nMJ:-Kn1uSp9mOSKU4Rq2nMK:b-4iv64e) ([raw file](https://github.com/zanata/zanata-platform/blob/e753e86ce8476996501daa1587731bee69e8a585/server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java#L82)):*
> ```Java
>             for (Map.Entry<String, AsyncTaskHandle<?>> entry : handlesByKey
>                     .entrySet()) {
>                 if (entry.getValue().equals(taskHandle)) {
> ```

We should change AsyncTaskHandle to hold its Key (or the String), so that we can extract it and use it for efficient lookups (O(1) instead of O(n)).

---

*[server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java, line 138 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kn1uwxQo0FRuaOjIacg:-Kn1uwxQo0FRuaOjIach:bschxed) ([raw file](https://github.com/zanata/zanata-platform/blob/e753e86ce8476996501daa1587731bee69e8a585/server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java#L138)):*
> ```Java
>          * @return String representation of the key which can be used as id
>          */
>         default String joinFields(String keyName, String... fields) {
> ```

This might be better as `static` than `default`.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 202 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmynr2BhPu3cN8SJEzQ:-Kn1yX0mysjQ-xyt9QRH:b-ljrjgd) ([raw file](https://github.com/zanata/zanata-platform/blob/ddc82d3fc6a6c5c60144faa09e7fc5243bdbf23e/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L202)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

There are some handles that just call the cancel method without setting CancelledBy and CancelledTime. I tried not to make too big a change to the existing code.
</blockquote></details>

That sounds like kind of a bug then, which crept in because the API didn't enforce it before. 

We could always have two cancel methods, one of which doesn't require a username but is deprecated, to be called by the old bad code.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 126 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kn1vDKksifVQ8XTKnBL:-Kn1vDKlp0CoS1AihuiX:bmte5xy) ([raw file](https://github.com/zanata/zanata-platform/blob/e753e86ce8476996501daa1587731bee69e8a585/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L126)):*
> ```Java
>                 String triggeredBy =
>                         ((UserTriggeredTaskHandle) handle).getTriggeredBy();
>                 if (!triggeredBy.equals(identity.getAccountUsername())) {
> ```

If you move this check into, say, `UserTriggeredTaskHandle.visibleTo(String accountName)`, we should be able to change the access rules more cleanly, and share the logic in non-REST parts of the system.

---

*[server/zanata-war/src/test/java/org/zanata/rest/service/AsyncProcessServiceTest.java, line 131 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-KmytegQ8RDNZfZ9SOSH:-Kn1zJ8Kxc1Q25ptREXK:b4n0k3l) ([raw file](https://github.com/zanata/zanata-platform/blob/ddc82d3fc6a6c5c60144faa09e7fc5243bdbf23e/server/zanata-war/src/test/java/org/zanata/rest/service/AsyncProcessServiceTest.java#L131)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

This is controlled by CDI interceptor. I added an arquillian test for this.
</blockquote></details>

Okay. We should have some comments explaining why those tests aren't here in this class.

---

*[server/zanata-war/src/test/java/org/zanata/rest/service/AsyncProcessServiceTest.java, line 111 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kn1vwfbSFXNoSCCwClC:-Kn1vwfbSFXNoSCCwClD:b9ontk8) ([raw file](https://github.com/zanata/zanata-platform/blob/e753e86ce8476996501daa1587731bee69e8a585/server/zanata-war/src/test/java/org/zanata/rest/service/AsyncProcessServiceTest.java#L111)):*
> ```Java
> 
>     @Test
>     public void adminUserCanNotSeeOthersProcessStatus() throws URISyntaxException {
> ```

"Not" doesn't belong in the name.

---

*[server/zanata-war/src/test/java/org/zanata/rest/service/raw/AsyncProcessRawRestITCase.java, line 66 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kn1xAjXInWmIvehqh2B:-Kn1xAjXInWmIvehqh2C:badmvgd) ([raw file](https://github.com/zanata/zanata-platform/blob/e753e86ce8476996501daa1587731bee69e8a585/server/zanata-war/src/test/java/org/zanata/rest/service/raw/AsyncProcessRawRestITCase.java#L66)):*
> ```Java
>     @Test
>     @RunAsClient
>     public void adminCanGetAllProcessStatuses() throws Exception {
> ```

Doesn't this duplicate one of the tests in AsyncProcessServiceTest?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/398)*
<!-- Sent from Reviewable.io -->

huangp commented:
Review status: 16 of 30 files reviewed at latest revision, 9 unresolved discussions.

---

*[server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java, line 62 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kn1rMojkpDZMr806pgg:-Kn2589KqP5HI-sxR6IM:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/e753e86ce8476996501daa1587731bee69e8a585/server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java#L62)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

This could use HashMap.putIfAbsent now, instead of containsKey/put.
</blockquote></details>

Done.

---

*[server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java, line 72 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kn1s9xbDtOjGQM2qVoM:-Kn259s9qZxzv_6FQETa:b-t8abe8) ([raw file](https://github.com/zanata/zanata-platform/blob/e753e86ce8476996501daa1587731bee69e8a585/server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java#L72)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

So we're using AsyncTaskKey like a FunctionalInterface (even though it doesn't have the annotation)? This seems error-prone, because the temptation in this method would be to write:

    AsyncTaskKey autoKey = () -> UUID.randomUUID().toString();

Whereas one of the most important things about a key is that it be immutable.

If it's going to be a FunctionalInterface, perhaps it would be better as an AsyncTaskKeyGen (after all, it just generates the id string which is the real key now).

Otherwise it might be safest if AsyncTaskKey were a final immutable class which just holds a String.

</blockquote></details>

No. We have a bunch of subclasses implementing their own id() method. Those keys can have more semantic meanings of what it represent.  This is the default one where you just want a unique random key (e.g. allowing the same type of process to run at the same time). I will create a GenericKey as subclass so that it's immutable.

---

*[server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java, line 82 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kn1uSp9mOSKU4Rq2nMJ:-Kn27mTzrzm-2MJvvp3e:b-6wcf0q) ([raw file](https://github.com/zanata/zanata-platform/blob/e753e86ce8476996501daa1587731bee69e8a585/server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java#L82)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

We should change AsyncTaskHandle to hold its Key (or the String), so that we can extract it and use it for efficient lookups (O(1) instead of O(n)).
</blockquote></details>

Thought about that. I wanted to do incremental change and merge so that this PR won't grow too long and stale. This PR is intended just to fix the Async key and add a new REST api for general control of tasks. Other refactoring or improvement of existing code can come later (this change will touch multiple JSF action class and a few other REST endpoints).

---

*[server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java, line 138 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kn1uwxQo0FRuaOjIacg:-Kn28apwvnXB5qNrzEgQ:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/e753e86ce8476996501daa1587731bee69e8a585/server/zanata-war/src/main/java/org/zanata/async/AsyncTaskHandleManager.java#L138)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

This might be better as `static` than `default`.
</blockquote></details>

Done.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 202 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kmynr2BhPu3cN8SJEzQ:-Kn2IAAX3NHOzz4dK4m9:bqdkzfo) ([raw file](https://github.com/zanata/zanata-platform/blob/ddc82d3fc6a6c5c60144faa09e7fc5243bdbf23e/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L202)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

That sounds like kind of a bug then, which crept in because the API didn't enforce it before. 

We could always have two cancel methods, one of which doesn't require a username but is deprecated, to be called by the old bad code.
</blockquote></details>

Again as discussed I want to merge often and sooner so not to change too much of existing code. I need to study for each individual task, how the cancellation is triggered (by user or by system) and how those values (cancelledBy and time) are being used. This can come as a separate PR.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java, line 126 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kn1vDKksifVQ8XTKnBL:-Kn2I7J_wQZrDGYd4ZKb:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/e753e86ce8476996501daa1587731bee69e8a585/server/zanata-war/src/main/java/org/zanata/rest/service/AsyncProcessService.java#L126)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

If you move this check into, say, `UserTriggeredTaskHandle.visibleTo(String accountName)`, we should be able to change the access rules more cleanly, and share the logic in non-REST parts of the system.
</blockquote></details>

Done.

---

*[server/zanata-war/src/test/java/org/zanata/rest/service/AsyncProcessServiceTest.java, line 131 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-KmytegQ8RDNZfZ9SOSH:-Kn24-mM-R6Dzay6mjD-:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/ddc82d3fc6a6c5c60144faa09e7fc5243bdbf23e/server/zanata-war/src/test/java/org/zanata/rest/service/AsyncProcessServiceTest.java#L131)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Okay. We should have some comments explaining why those tests aren't here in this class.
</blockquote></details>

Done.

---

*[server/zanata-war/src/test/java/org/zanata/rest/service/AsyncProcessServiceTest.java, line 111 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kn1vwfbSFXNoSCCwClC:-Kn240sdwmYpy981AAoA:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/e753e86ce8476996501daa1587731bee69e8a585/server/zanata-war/src/test/java/org/zanata/rest/service/AsyncProcessServiceTest.java#L111)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

"Not" doesn't belong in the name.
</blockquote></details>

Done.

---

*[server/zanata-war/src/test/java/org/zanata/rest/service/raw/AsyncProcessRawRestITCase.java, line 66 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-Kn1xAjXInWmIvehqh2B:-Kn242DMu0fOjswYpX9o:b-6714fz) ([raw file](https://github.com/zanata/zanata-platform/blob/e753e86ce8476996501daa1587731bee69e8a585/server/zanata-war/src/test/java/org/zanata/rest/service/raw/AsyncProcessRawRestITCase.java#L66)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Doesn't this duplicate one of the tests in AsyncProcessServiceTest?
</blockquote></details>

This checks the admin permission (go through CDI interceptor). It doesn't set up any real data (the result is empty since nothing is running or finished on server). The one in unit test checks the actual implementation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/398)*
<!-- Sent from Reviewable.io -->

seanf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 18 files at r5, 14 of 14 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/398#-:-Kn33R_pdklgkJJBKcgc:bnfp4nl)*
<!-- Sent from Reviewable.io -->

djansen-redhat commented:
Reviewed 10 of 27 files at r1, 7 of 18 files at r5, 12 of 14 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/398)*
<!-- Sent from Reviewable.io -->
